external/avro/src/main/scala/org/apache/spark/sql/avro/SparkAvroDatumReader.scala

```diff
@@ -0,0 +1,811 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.avro
+
+import java.io.IOException
+import java.math.BigDecimal
+import java.nio.ByteBuffer
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.avro.{LogicalTypes, Schema, SchemaBuilder}
+import org.apache.avro.Conversions.DecimalConversion
+import org.apache.avro.LogicalTypes.{TimestampMicros, TimestampMillis}
+import org.apache.avro.Schema.Type._
+import org.apache.avro.generic.{GenericDatumReader, GenericFixed}
+import org.apache.avro.io.ResolvingDecoder
+import org.apache.avro.util.Utf8
+
+import org.apache.spark.sql.catalyst.{InternalRow, NoopFilters, StructFilters}
+import org.apache.spark.sql.catalyst.expressions.{MutableAny, MutableBoolean, MutableByte,
+  MutableDouble, MutableFloat, MutableInt, MutableLong, MutableShort, MutableValue,
+  SpecificInternalRow, UnsafeArrayData}
+import org.apache.spark.sql.catalyst.util.{ArrayBasedMapData, ArrayData, DateTimeUtils,
+  GenericArrayData}
+import org.apache.spark.sql.catalyst.util.DateTimeConstants.MILLIS_PER_DAY
+import org.apache.spark.sql.execution.datasources.DataSourceUtils
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.internal.SQLConf.LegacyBehaviorPolicy
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+/**
+ * Reader to read data as catalyst format from avro decoders. Multiple calls
+ * return the same [[InternalRow]] object. If needed, caller should copy before making a new call.
+ *
+ * @param actual
+ * @param expected
+ * @param rootCatalystType
+ * @param datetimeRebaseMode
+ * @tparam T
+ */
+class SparkAvroDatumReader[T](
```

Comment:
This commit moves to using native reader for avro files. Previously we had Decoder => GenericAvroDatumReader => AvroDeserializer. Now SparkAvroDatumReader reads directly from the Decoder. Array reading performance degraded, which has been fixed in a latter commit. Moving to the native reader also allows us to "skip" reading columns that are not needed. This change is also in a latter commit.

Read times can be found in the attached pdf on pg 2 under column D.